### DISCUSSION TITLE
How to use ImageMagick to add line numbers to scanned documents?

### DISCUSSION THREAD

**Discussion by tommydog35** [ImageMagick/ImageMagick#3102] [Help]
I am trying to accomplish the following and wondered if ImageMagick could provide a solution:

1) I have many scanned images of text documents and need to add line numbers to every line. 

2)  I also have thousands of pages of PDF documents which are not scanned, however I don't have the source text and need to add line numbers to these.  It seems that someone has posted a solution to this here:

https://tex.stackexchange.com/questions/18760/add-page-and-line-numbers-to-a-pdf/18776#18776

However I really don't understand the above explanation.

If anyone has a simple explanation as to how I could achieve any of the above (solving the PDF problem would save me the most work) in ImageMagick, I would be very grateful.  My knowledge of ImageMagick is very limited and it's the first command line programme I have ever used!  I am using Windows, but if this would be easier to accomplish in Linux, I could use Ubuntu.

I could obviously do this manually, but it would be extremely tedious over thousands of pages. In addition, I don't want to change the formatting of the documents, as they are legal documents.

**Comment by snibgo** [comment]
> However I really don't understand the above explanation.
> If anyone has a simple explanation as to how I could achieve any of the above...

The link shows how to solve the problem using three tools: ImageMagick and sed to find where the text lines are, and LaTeX to build a document containing the line numbers. I know nothing about LaTeX.

The link uses ImageMagick to rasterize PDF pages. If the PDF files are simply raster scans, then a more direct method that gives better quality is to extract the scans with pdfimages.

A solution could then use just ImageMagick to do both parts of the task. However, the output would be raster images.

**Reply by tommydog35** [reply]
> A solution could then use just ImageMagick to do both parts of the task. However, the output would be raster images.

Supposing I have extracted the images and have a folder of images, how can I use ImageMagick to number the lines?  I don't mind the output being raster images.

**Comment by snibgo** [comment]
For each file, when IM and sed have created a list of y-coordinates, a script would step through that list, writing line numbers at those coordinates.

**Reply by tommydog35** [reply]
> For each file, when IM and sed have created a list of y-coordinates, a script would step through that list, writing line numbers at those coordinates.

Sorry but I have no idea how to do this.  Are there any examples or simple tutorials which show how to do this, as my knowledge is zero.

**Comment by snibgo** [comment]
Please link to a sample input PDF.

**Reply by tommydog35** [reply]
> Please link to a sample input PDF.

Below is a link to a sample PDF.  The sample reflects the fact that all pages are often very different. 

https://pdfhost.io/v/MVXcZyr51_samplepdf.pdf

**Comment by snibgo** [comment]
The following copes with pages that have only text (not images) that is not in columns. Windows BAT syntax:
```
set LineNum=1

for /F "usebackq tokens=2 delims=," %%A in (`magick ^
  sample3.pdf[7] ^
  -background White -layers Flatten ^
  -alpha off ^
  +write page.miff ^
  -set option:HH %%[fx:h] ^
  +write x0.png ^
  -morphology open rectangle:10x2 ^
  -morphology open rectangle:1x5 ^
  +write x1.png ^
  -scale "1x^!" ^
  -threshold 97%% ^
  +write x2.png ^
  -morphology HMT "1x2:1,0" ^
  -transparent Black ^
  +write x3.png ^
  -write sparse-color: ^
  +delete ^
  -size 50x%%[HH] xc: ^
  +write linenums.miff ^
  NULL: ^|sed -e 's/ /\n/g' `) do (
  echo %%A
  %IMG7%magick linenums.miff -gravity NorthEast -annotate +0+%%A !LineNum! linenums.miff
  set /A LineNum+=1
)

magick ^
  linenums.miff page.miff ^
  +append ^
  out.png
```
The script could be called for all pages in a PDF, and all PDFs in a folder.
![out](https://user-images.githubusercontent.com/33812368/104112495-8ce9fd00-52e7-11eb-92f6-8969b400659f.png)

**Reply by tommydog35** [reply]
> The following copes with pages that have only text (not images) that is not in columns. Windows BAT syntax:
> 
> ```
> set LineNum=1
> 
> for /F "usebackq tokens=2 delims=," %%A in (`magick ^
>   sample3.pdf[7] ^
>   -background White -layers Flatten ^
>   -alpha off ^
>   +write page.miff ^
>   -set option:HH %%[fx:h] ^
>   +write x0.png ^
>   -morphology open rectangle:10x2 ^
>   -morphology open rectangle:1x5 ^
>   +write x1.png ^
>   -scale "1x^!" ^
>   -threshold 97%% ^
>   +write x2.png ^
>   -morphology HMT "1x2:1,0" ^
>   -transparent Black ^
>   +write x3.png ^
>   -write sparse-color: ^
>   +delete ^
>   -size 50x%%[HH] xc: ^
>   +write linenums.miff ^
>   NULL: ^|sed -e 's/ /\n/g' `) do (
>   echo %%A
>   %IMG7%magick linenums.miff -gravity NorthEast -annotate +0+%%A !LineNum! linenums.miff
>   set /A LineNum+=1
> )
> 
> magick ^
>   linenums.miff page.miff ^
>   +append ^
>   out.png
> ```
> 
> The script could be called for all pages in a PDF, and all PDFs in a folder.



Thank you so much for providing this.  I copied the code to notepad and saved it as lines.BAT, however when I run  lines.BAT, I get the following error in CMD when I try and apply it to the same example that you posted:

The system cannot find message text for message number 0x2350 in the message fil
e for Application.

Copyright (c) 2009 Microsoft Corporation.  All rights reserved.
Not enough storage is available to process this command.

C:\Program Files\ImageMagick-7.0.3-Q16>lines.bat

C:\Program Files\ImageMagick-7.0.3-Q16>set LineNum=1

C:\Program Files\ImageMagick-7.0.3-Q16>for /F "usebackq tokens=2 delims=," %A in
 (`magick   sample3.pdf[7]   -background White -layers Flatten   -alpha off   +w
rite page.miff   -set option:HH %[fx:h]   +write x0.png   -morphology open recta
ngle:10x2   -morphology open rectangle:1x5   +write x1.png   -scale "1x^!"   -th
reshold 97%   +write x2.png   -morphology HMT "1x2:1,0"   -transparent Black   +
write x3.png   -write sparse-color:   +delete   -size 50x%[HH] xc:   +write line
nums.miff   NULL: |sed -e 's/ /\n/g' `) do (
echo %A
 magick linenums.miff -gravity NorthEast -annotate +0+%A !LineNum! linenums.miff

 set /A LineNum+=1
)
sed: -e expression #1, char 1: unknown command: `''
magick: FailedToExecuteCommand `"gswin32c.exe" -q -dQUIET -dSAFER -dBATCH -dNOPA
USE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 "-sDEVICE=p
ngalpha" -dTextAlphaBits=4 -dGraphicsAlphaBits=4 "-r72x72" -dFirstPage=8 -dLastP
age=8 "-sOutputFile=C:/Users/joe/AppData/Local/Temp/magick-6544ZAgnBnBJNk5g%d" "
-fC:/Users/joe/AppData/Local/Temp/magick-6544-2SwJAJEF7oL" "-fC:/Users/joe/AppDa
ta/Local/Temp/magick-6544gGrqRJkKYBHw"' (The system cannot find the file specifi
ed.
) @ error/delegate.c/ExternalDelegateCommand/460.
magick: PDFDelegateFailed `The system cannot find the file specified.
' @ error/pdf.c/ReadPDFImage/783.
magick: no image to apply a property "%w" @ warning/property.c/GetMagickProperty
Letter/2550.
magick: unknown image property "%w" @ warning/property.c/InterpretImagePropertie
s/3488.
magick: no image to apply a property "%h" @ warning/property.c/GetMagickProperty
Letter/2438.
magick: unknown image property "%h" @ warning/property.c/InterpretImagePropertie
s/3488.
magick: no image to apply a property "%m" @ warning/property.c/GetMagickProperty
Letter/2469.
magick: unknown image property "%m" @ warning/property.c/InterpretImagePropertie
s/3488.

C:\Program Files\ImageMagick-7.0.3-Q16>magick   linenums.miff page.miff   +appen
d   out.png
magick: unable to open image 'linenums.miff': No such file or directory @ error/
blob.c/OpenBlob/2695.
magick: no image to apply a property "%w" @ warning/property.c/GetMagickProperty
Letter/2550.
magick: unknown image property "%w" @ warning/property.c/InterpretImagePropertie
s/3488.
magick: no image to apply a property "%h" @ warning/property.c/GetMagickProperty
Letter/2438.
magick: unknown image property "%h" @ warning/property.c/InterpretImagePropertie
s/3488.
magick: no image to apply a property "%m" @ warning/property.c/GetMagickProperty
Letter/2469.
magick: unknown image property "%m" @ warning/property.c/InterpretImagePropertie
s/3488.

C:\Program Files\ImageMagick-7.0.3-Q16>



I would be really grateful if you could tell me how to get the script working.  Thank you so much.

**Comment by snibgo** [comment]
> magick: FailedToExecuteCommand "gswin32c.exe"

This means Ghostscript is not installed on your computer. IM needs Ghostscript to rasterize PDF files. You should download GS from https://ghostscript.com/ and install it. You may then need to re-install IM. As your version is old (v7.0.3), I suggest you downlod a more recent version and install that.

You can test whether IM and GS are working with:
```
magick sample3.pdf[7] x.png
```
Check for error mesaages.

**Reply by tommydog35** [reply]
@snibgo  means Ghostscript is not installed on your computer. IM needs Ghostscript to rasterize PDF files. You should download GS from https://ghostscript.com/ and install it. You may then need to re-install IM. As your version is old (v7.0.3), I suggest you downlod a more recent version and install that.
> 
> You can test whether IM and GS are working with:
>  magick sample3.pdf[7] x.png
> ```

I have done what you said and downloaded the latest edition of ImageMagick and Ghostscript.  I tested whether IM nad GS are working by magick sample3.pdf[7] x.png  and confirm that it is working.  But I am still having problems with the line number script.  When I run the script, the below files are generated along with linenums.miff and page.miff , but there are still no line numbers shown.  Below is the output of the CMD window when I run your script, along with the images generated:

The system cannot find message text for message number 0x2350 in the message fil
e for Application.

Copyright (c) 2009 Microsoft Corporation.  All rights reserved.
Not enough storage is available to process this command.

C:\Users\joe\Desktop\imagemagick\test>lines.bat

C:\Users\joe\Desktop\imagemagick\test>set LineNum=1

C:\Users\joe\Desktop\imagemagick\test>for /F "usebackq tokens=2 delims=," %A in
(`magick   sample3.pdf   -background White -layers Flatten   -alpha off   +write
 page.miff   -set option:HH %[fx:h]   +write x0.png   -morphology open rectangle
:10x2   -morphology open rectangle:1x5   +write x1.png   -scale "1x^!"   -thresh
old 97%   +write x2.png   -morphology HMT "1x2:1,0"   -transparent Black   +writ
e x3.png   -write sparse-color:   +delete   -size 50x%[HH] xc:   +write linenums
.miff   NULL: |sed -e 's/ /\n/g' `) do (
echo %A
 magick linenums.miff -gravity NorthEast -annotate +0+%A !LineNum! linenums.miff

 set /A LineNum+=1
)
sed: -e expression #1, char 1: unknown command: `''

C:\Users\joe\Desktop\imagemagick\test>magick   linenums.miff page.miff   +append
   out.png

C:\Users\joe\Desktop\imagemagick\test>

![x0](https://user-images.githubusercontent.com/77201980/104130209-6a97c400-5367-11eb-8f1c-cdb245e74cb3.png)
![x1](https://user-images.githubusercontent.com/77201980/104130217-71263b80-5367-11eb-82a7-a3d8500d006b.png)

![x2](https://user-images.githubusercontent.com/77201980/104130233-7b483a00-5367-11eb-9348-08e8c815e6fb.png)

![x3](https://user-images.githubusercontent.com/77201980/104130238-813e1b00-5367-11eb-9868-69b816764140.png)

**Comment by fmw42** [comment]
How many pages are in your PDF file. 

`magick sample3.pdf[7] x.png `

[7] means you are try to access the 8th page of the PDF. Does your PDF file have at least 8 pages?

**Reply by tommydog35** [reply]
> How many pages are in your PDF file.
> 
> `magick sample3.pdf[7] x.png `
> 
> [7] means you are try to access the 8th page of the PDF. Does your PDF file have at least 8 pages?

Yes I understand that and IM and GS are working now that I have downloaded the latest versions.  The problem I am having is to get the script that snibgo posted working.  It seems to generate a number of files, but not sure how these are used to number the lines?  Please see my previous post above.

**Reply by fmw42** [reply]
But you did not answer my question about how many pages in your PDF

**Comment by snibgo** [comment]
> The system cannot find message text for message number 0x2350 in the message file for Application.
> Copyright (c) 2009 Microsoft Corporation. All rights reserved.
> Not enough storage is available to process this command.

I don't know what software is giving those messages, or what they mean. A web search suggests a problem with Microsoft cmd.exe, that prevents error messages being properly displayed:

https://answers.microsoft.com/en-us/windows/forum/all/command-prompt-showing-0x2350-updates-installation/4dbc78fb-ceaf-46bd-a627-5ebb38bc6fec

https://domainwebcenter.com/resolved-the-system-cannot-find-message-text-for-message/

> sed: -e expression #1, char 1: unknown command: `''

Do you have the "sed" program? What happens when you type "sed" at the command line? Expect some help text, starting something like this:
```
Usage: sed [OPTION]... {script-only-if-no-other-script} [input-file]...

  -n, --quiet, --silent
                 suppress automatic printing of pattern space
  -e script, --expression=script
                 add the script to the commands to be executed
```

**Comment by fmw42** [comment]
Which convert are you using for you IM processing? Might you be using the Windows convert.exe rather than the IM convert.exe?

**Reply by snibgo** [reply]
@fmw42: That isn't the problem. IM "magick" is clearly working, and succesfully calling Ghostscript, as seen from the images made.

**Reply by tommydog35** [reply]
After further examination, I think the problem is with SED.  It seems that SED is not installed at the system level.  As when I run the script, I get "sed is not recognized as an internal or external command" I then went to the where sed is installed C:\Program Files (x86)\GnuWin32 and copied the sed files from the bin to the directory where the ImageMagick script is.  I then run the script, but it does not result in numbered lines, just the images created.  So I think the problem is with SED on windows.  Not sure how other people got SED to work properly on windows?

**Comment by snibgo** [comment]
We are moving forwards. "sed" works fine for me, on Windows 8.1. I installed sed and many other tools from Cygwin. Here, we are using sed to translate each space into a newline. Please try the following command:
```
echo first second third |sed -e 's/ /\n/g'
```

The expected output is:
```
first
second
third
```
EDIT: Incidentally, if "sed" doesn't work, we can use "tr" instead. Try:
```
echo first second third |tr ' ' \n
```

**Reply by tommydog35** [reply]
@snibgo >We are moving forwards. "sed" works fine for me, on Windows 8.1. I installed sed and many other tools from Cygwin. 

I installed Cygwin, which installed sed, but sed was still not available as a global command from cmd.  Furthermore, I found out I can't execute windows bat scripts from the Cygwin64 terminal.   So I  copied from the Cygwin64 installation folder the following files to the folder where your bat script is:
cygiconv-2.dll
cygintl-8.dll
cygwin1.dll
sed.exe

Good progress has now been made, as your bat script now runs without errors and has produced the attached image (out.pgn) below.  Your script appears to have correctly identified the lines.  The script has printed at the start of every line the words "LineNum!"  but the problem is it has not printed the actual numbers, just the words "LineNum!"  How do I fix this to show actual line numbers?

![out](https://user-images.githubusercontent.com/77201980/104139190-8e77fb80-53a1-11eb-83d8-12e9cd84b892.png)

**Comment by snibgo** [comment]
> ... but sed was still not available as a global command from cmd.

I suppose the directory containing sed.exe isn't on your system path. I suggest you add it to your path.

For !LineNum!, insert the following line at the top of the script:
```
setlocal enabledelayedexpansion
```

**Reply by tommydog35** [reply]
Thank you snibgo very much for helping.  With your help I now have it working as it should - thank you. I rally appreciate it.

 Just one question - what needs to be changed in your script to get the script to number lines for all pages in a PDF?  In your script you used [7] to denote the page for processing.  But I want the script to write line numbers to every page in a PDF.

**Comment by snibgo** [comment]
Whew. I'm glad it works.

Okay, now it works for one page of one PDF. I suggest you change it so it takes three parameters, for the input filename, page number and output file. Like this:
```
rem %1 is input PDF file
rem %2 is a page number within that PDF
rem %3 is output filename

setlocal enabledelayedexpansion

set LineNum=1

for /F "usebackq tokens=2 delims=," %%A in (`magick ^
  %1[%2] ^
  -background White -layers Flatten ^
  -alpha off ^
  +write page.miff ^
  -set option:HH %%[fx:h] ^
  +write x0.png ^
  -morphology open rectangle:10x2 ^
  -morphology open rectangle:1x5 ^
  +write x1.png ^
  -scale "1x^!" ^
  -threshold 97%% ^
  +write x2.png ^
  -morphology HMT "1x2:1,0" ^
  -transparent Black ^
  +write x3.png ^
  -write sparse-color: ^
  +delete ^
  -size 50x%%[HH] xc: ^
  +write linenums.miff ^
  NULL: ^|sed -e 's/ /\n/g' `) do (
  echo %%A
  magick linenums.miff -gravity NorthEast -annotate +0+%%A !LineNum! linenums.miff
  set /A LineNum+=1
)

magick ^
  linenums.miff page.miff ^
  +append ^
  %3
```

Suppose that script is called "pdfLineNums.bat". Then you can call it like this:
```
call pdfLineNums sample3.pdf 7 sample3_7.tiff
```
If you want to loop through all the pages within one PDF, write another script called onePdfLines.bat:
```
echo Processing File %1

for /F "usebackq" %%L in (`magick -ping %1 -format "NumFrames=%%N\n" info:`) do set %%L

set /A Last=%NumFrames%-1

echo Last=%Last%

for /L %%I in (0,1,%Last%) do call pdfLineNums %1 %%I %1_%%I.tiff
```
Then we can run this command:
```
call onePdfLines.bat sample3.pdf
```
... which creates files called sample3_0.tiff, sample3_1.tiff ... sample3_7.tiff.

We can create another script that loops through all the PDF files in a directory, and calls onePdfLines.bat for each PDF.

**Comment by snibgo** [comment]
You may have two more questions. I will give outline answers.


Q1. What about PDF pages that have embedded raster images, such as "sample3.pdf[0]"?

https://ghostscript.com/doc/9.53.3/Use.htm#Options says:
```
-dFILTERIMAGE
    If set, this will ignore all images in the input (in this context image means a bitmap), these will therefore not be rendered. 
```
So we can include this before rasterising:
```
rem If PDF contains raster images, this will suppress them, rasterising just text.
set GS_OPTIONS=-dFILTERIMAGE
```
But if the PDF page is a scanned document, we won't get that scanned image, so we won't get any text. So we probably need to rasterize with that option then, if the image is just white, without that option.

If the PDF contains an embedded raster such as a face, and scanned text as a raster, we would ideally remove the face but not the text. That task is not simple.


Q2. What about PDF pages with two columns of text? Can we number lines on the left of the left half, then on the right of the right half?

First, see if the page has a wide gutter of white down the middle. If it does, then chop the page in half, get the line numbers for the left in one image, and the line numbers for the right in another image, then append all four images together. This could be extended for any number of columns.

**Reply by tommydog35** [reply]
Thank you snibgo for your very clear instructions.  I have followed what you said and used onePdfLines.bat to loop through all the pages within a single PDF.  This works, but the problem is each of the output tiff files (0.tiff   1.tiff   2.tiff etc) starts numbering from 1 on every page.  I need continuous numbering, so if 0.tiff ends with number 40, then 1.tiff should start with number 41 and so on.  How can the script be changed to achieve continuous numbering?

**Comment by snibgo** [comment]
Okay, we want to start LineNum=1 at each new PDF instead of each page. That is a question about scripting, rather than ImageMagick. There are many possible solutions, and I show one below.

This also shows a method for suppressing embedded raster images, provided this doesn't make totally white pages.


pdfLineNums.bat:
```
rem %1 is input PDF file
rem %2 is a page number within that PDF
rem %3 is output filename
rem %4 is first line number

setlocal enabledelayedexpansion

set LineNum=%4
if "%LineNum%"=="" set LineNum=1

set GS_OPTIONS=

rem If PDF contains images, this will supress them, rasterising just text.
set GS_OPTIONS=-dFILTERIMAGE

for /F "usebackq" %%L in (`magick ^
  %1[%2] ^
  -background White -layers Flatten ^
  -alpha off ^
  +write page.miff ^
  -format "MINVAL=%%[fx:minima]\n" ^
  info:`) do set %%L

set GS_OPTIONS=

if "%MINVAL%"=="1" magick ^
  %1[%2] ^
  -background White -layers Flatten ^
  -alpha off ^
  page.miff

for /F "usebackq tokens=2 delims=," %%A in (`magick ^
  page.miff ^
  -set option:HH %%[fx:h] ^
  +write x0.png ^
  -morphology open rectangle:10x2 ^
  -morphology open rectangle:1x5 ^
  +write x1.png ^
  -scale "1x^!" ^
  -threshold 97%% ^
  +write x2.png ^
  -morphology HMT "1x2:1,0" ^
  -transparent Black ^
  +write x3.png ^
  -write sparse-color: ^
  +delete ^
  -size 50x%%[HH] xc: ^
  +write linenums.miff ^
  NULL: ^|sed -e 's/ /\n/g' `) do (
  echo LineNum=!LineNum! at y=%%A
  magick linenums.miff -gravity NorthEast -annotate +0+%%A !LineNum! linenums.miff
  set /A LineNum+=1
)

magick ^
  linenums.miff page.miff ^
  +append ^
  %3

endlocal & set LineNum=%LineNum%
```

onePdfLines.bat:
```
setlocal enabledelayedexpansion

echo %0: Processing File %1

for /F "usebackq" %%L in (`magick -ping %1 -format "NumFrames=%%N\n" info:`) do set %%L

set /A Last=%NumFrames%-1

echo Last=%Last%

set LineNum=1

for /L %%I in (0,1,%Last%) do call pdfLineNums %1 %%I %~n1_%%I.tiff !LineNum!
```

**Reply by tommydog35** [reply]
Thank you snibgo for tweaking the script for continuous numbering - it works perfectly.  I also see you have modified the script to add GS_OPTIONS=-dFILTERIMAGE.  The problem is GS_OPTIONS=-dFILTERIMAGE resulted in  removing pictures from the pages.  But what if I don't want to remove the pictures and just ignore the pictures so that they are not numbered at all?  Is there a basic way to do this, where pictures are just treated like blank lines and ignored in the numbering?

**Comment by snibgo** [comment]
The final command appends the image of linenumbers with page.miff.
```
magick ^
  linenums.miff page.miff ^
  +append ^
  %3
```
You can append it to the original page if you prefer:
```
magick ^
  linenums.miff %1[%2] ^
  +append ^
  %3
```

**Reply by tommydog35** [reply]
Well if I use the below code, it looks strange.  I think I may just remove the dFILTERIMAGE, as I can't see the advantage?  
magick ^
  linenums.miff %1[%2] ^
  +append ^
  %3

Also when including the above code, it produces a strange output with a white column of numbers on the page, and seems to have no advantage when dealing with pictures on the page.  Or maybe I misunderstood something?  Please see the images below:

![sample 0](https://user-images.githubusercontent.com/77201980/104248454-6d76df80-5461-11eb-97e4-0531e18f9700.png)
![sample 1](https://user-images.githubusercontent.com/77201980/104248462-736cc080-5461-11eb-8be0-60977fb7568f.png)

On a separate point, are there any values in your script that can be changed to determine final output quality?  If I am not worried about file size and my only goal is to have final images that are the best quality for printing on A4 paper?  I know there will be a quality loss when extracting from a pdf and making raster images for the output, but what is the best way of preserving as much quality as possible?

**Comment by snibgo** [comment]
When rasterizing a PDF, you can specify a "-density" (in pixels per inch) before reading the PDF. Default is 72. Higher numbers (eg 150, 300, 600) give better quality but more pixels so take longer to process and make larger files. For example:
```
magick -density 150 sample3.pdf[7] info:
```

**Reply by tommydog35** [reply]
> When rasterizing a PDF, you can specify a "-density" (in pixels per inch) before reading the PDF. Default is 72. Higher numbers (eg 150, 300, 600) give better quality but more pixels so take longer to process and make larger files. For example:
> 
> ```
> magick -density 150 sample3.pdf[7] info:
> ```

Thanks for the tip, but how do I specify this when calling the script?  I have tried doing it like this:

call onePdfLines.bat -density 300 sample3.pdf

But it does not work.

**Comment by snibgo** [comment]
You need to change the script from ...
```
for /F "usebackq" %%L in (`magick ^
  %1[%2] ^
```
... to ...
```
for /F "usebackq" %%L in (`magick ^
  -density 300 ^
  %1[%2] ^
```

**Reply by tommydog35** [reply]
Many thanks Snibgo.  I changed the script as you said and it works to produce higher quality files, however if I choose a higher value e.g. 300, the line numbers become very small on the page (almost unreadable) and don't always appear inline with the text.  If the density is changed to a higher number, is there a way to scale the line number text accordingly?

**Comment by snibgo** [comment]
The default density is 72. If you use "-density 300", this is about 4 times as large.

The line numbers are written with "-annotate". This has a default pointsize of 12. You want it 4 times as large, so insert "-pointsize 48" just before "-annotate".

The image that is annotated is created by "-size 50x%%[HH] xc:", so it is 50 pixels wide. This may not be large enough for larger text, so also multiply this by 200, like this: "-size 200x%%[HH] xc:".

**Reply by tommydog35** [reply]
Thanks snibgo you have been really helpful.
